### PR TITLE
loading exons with coords behind length for circular seq_regions

### PR DESCRIPTION
**Pull request by vsitnik**
Fixing exon coordinates for models on circular seq_regions, with coordinates behind the seq_region length.
We need to
1) be able to store exon coordinates laying behind the circular seq_region end within seq_region (using "modular arithmetics")
2) be able to proper infer translation start/end in the cases when rank ordering is not the same as the adjusted  genome coordinates ordering (end is followed by start, followed by middle part) (or transpliced cases)
3) fix exon phases when `transcript::coding_region_start` invariant is broken on circular chromosomes  and `Exon::coding_region_start` is broken as well.
(2 & 3 are fixed by processing exons according to their ranks, default ordering returned by `Transcript::get_all_Exons`)

Tested on  `ensembl-production-metazoa/meta/plants/110/rs8t_gsoja` (GCF_004193775.1_ASM419377v2, Wild soybean) loading.

Trying to fix loading and proper phase inference for models like
https://github.com/Ensembl/ensembl-production-metazoa/blob/47c0d9b605df6019faa6a99f79675478f92ef750/meta/plants/110/rs8t_gsoja#L53 :
```
# dealing with partial transpliced genes "gene-W848_p084" and "gene-W848_p0845" on "NC_022868.1" (chloroplast)
# here are some intermixed gbff and gff3 lines
#
#LOCUS       NC_022868             152217 bp    DNA     circular PLN 19-NOV-2013
...
# "W848_p085"
#     gene            complement(join(96742..97530,68177..68302))
#                     /locus_tag="W848_p085"
#NC_022868.1  RefSeq  gene  68177  68302  .  -  .  ID=gene-W848_p085;Dbxref=GeneID:17675718
#NC_022868.1  RefSeq  gene  96742  97530  .  -  .  ID=gene-W848_p085;Dbxref=GeneID:17675718
#
#NC_022868.1  RefSeq  mRNA  96742  220519  .  -  .  ID=rna-W848_p085;Parent=gene-W848_p085
#NC_022868.1  RefSeq  exon  220394  220519  .  -  .  ID=exon-W848_p085-1;Parent=rna-W848_p085
#NC_022868.1  RefSeq  exon  97299  97530  .  -  .  ID=exon-W848_p085-2;Parent=rna-W848_p085
#NC_022868.1  RefSeq  exon  96742  96767  .  -  .  ID=exon-W848_p085-3;Parent=rna-W848_p085
#
#     CDS             complement(join(96742..96767,97299..97530,68177..68302))
#NC_022868.1  RefSeq  CDS  220394  220519  .  -  0  ID=cds-YP_008816226.1;Parent=rna-W848_p085
#NC_022868.1  RefSeq  CDS  97299  97530  .  -  0  ID=cds-YP_008816226.1;Parent=rna-W848_p085
#NC_022868.1  RefSeq  CDS  96742  96767  .  -  2  ID=cds-YP_008816226.1;Parent=rna-W848_p085
#
# For this one, let's rather preserve coordinates > contig_len
# drop gene part 1 (gene 68177 68302) and change part 2 (gene 96742 97530) to
# NC_022868.1  RefSeq  gene  96742  220519 .  -  .  ID=gene-W848_p085;Dbxref=GeneID:17675718
```

### CODE REVIEW

**Comment by vsitnik** [comment]
An example meta config file:
https://github.com/Ensembl/ensembl-production-metazoa/blob/main/meta/plants/110/rs8t_gsoja#L20

we have to alter gff3 file like this
https://github.com/Ensembl/ensembl-production-metazoa/blob/c2a147574a634a14fd09f65d2454cd222b8d27ae/meta/plants/110/rs8t_gsoja#L46
https://github.com/Ensembl/ensembl-production-metazoa/blob/c2a147574a634a14fd09f65d2454cd222b8d27ae/meta/plants/110/rs8t_gsoja#L78

And add this exons rank and phase fixes:
https://github.com/Ensembl/ensembl-production-metazoa/blob/c2a147574a634a14fd09f65d2454cd222b8d27ae/meta/plants/110/rs8t_gsoja#L193
```
# FIX W848_p084 / GeneID_17675718 related exons
# CDS             join(complement(68177..68302),137862..138093,138625..138650)
#NC_022868.1  RefSeq  CDS  68177  68302  .  -  0  ID=cds-YP_008816270.1;Parent=rna-W848_p084
#NC_022868.1  RefSeq  CDS  137862  138093  .  +  0  ID=cds-YP_008816270.1;Parent=rna-W848_p084
#NC_022868.1  RefSeq  CDS  138625  138650  .  +  2  ID=cds-YP_008816270.1;Parent=rna-W848_p084
# exon_id seq_region_id   seq_region_start        seq_region_end  seq_region_strand       phase   end_phase       is_current      is_constitutive stable_id       version created_datemodified_date    exon_id transcript_id   rank
# 11413   1120    138625  138650  1       0       2       1       0       W848_p084_t1-E1 NULL    <DATE>     <DATE>     11413   2760    1
# 11414   1120    68177   68302   -1      -1      -1      1       0       W848_p084_t1-E2 NULL    <DATE>     <DATE>     11414   2760    2
# 11415   1120    137862  138093  1       -1      -1      1       0       W848_p084_t1-E3 NULL    <DATE>     <DATE>     11415   2760    3
# change ranks (1,2,3) to (3,1,2):
# change phase to (2, 0, 0)
# $CMD_W -D prepl_glycine_soja_gca004193775v2rs_core_57_110_1 -e '
#   update exon_transcript set rank = 3 where exon_id = 11413;
#   update exon_transcript set rank = 1 where exon_id = 11414;
#   update exon set phase = 0, end_phase = 0 where exon_id = 11414;
#   update exon_transcript set rank = 2 where exon_id = 11415;
#   update exon set phase = 0, end_phase = 1 where exon_id = 11415;
#   update translation set  start_exon_id = 11414, end_exon_id = 11413, seq_end = 25 where translation_id = 2233;
#   update translation_attrib set value = replace(value, "50", "128")  where translation_id = 2233 and attrib_type_id  = 144;
# '
```

File: src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm [hunk 1]
```diff
@@ -398,20 +400,78 @@ sub add_pseudogenic_transcript {
...
+    # ideally, just call
+    #   $ga->update_coords($gene);
+    # unfortunately this doesn't work for circular seq_regions
+    # ( Gene::seq_region_start/end uses BaseFeatureAdaptor::_seq_region_boundary_from_db failback for circular regions )
+    # going vanila SQL update way
+    $self->update_gene_seq_region_coords_from_coords($gene, $ga);
+  }
+}
+
+sub update_gene_seq_region_coords_from_coords {
```
Comment (MatBarba) [inline src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm:423]: It's not clear what the circular case does differently from the default update_coords? Please add documentation

**Review by MatBarba — APPROVED** [review]
This is sound, just a few typos and some requests to add some documentation for key points.

File: src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm (hunk 1 above)
Comment (vsitnik) [reply src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm:423]: See on L414. should I copy it?

File: src/python/ensembl/brc4/runnable/integrity.py [hunk 2]
```diff
@@ -532,23 +534,36 @@ def check_seq_region_lengths(self, seqrs, feats, name):
...
+        diff_circular = []
+        diff_circular_list = []
+
         for seq_id in seqrs:
             if seq_id in feats:
                 # Check that feature is within the seq_region length
                 if feats[seq_id] > seqrs[seq_id]:
-                    diff.append(seq_id)
-                    diff_list.append("%s: %d vs %d" % (seq_id, seqrs[seq_id], feats[seq_id]))
+                      if circular is None or not circular.get(seq_id, False):
```
[the remark this answers is not in the document]
Comment (vsitnik) [reply src/python/ensembl/brc4/runnable/integrity.py:545]: if there's no dictionary or seq_region is not in the dictionary append it to the diff list. should be ok with `not
`

[the remark this answers is not in the document]
Comment (JAlvarezJarreta) [reply src/python/ensembl/brc4/runnable/integrity.py:545]:
```suggestion
                      if circular is None or seq_id not in circular:
```
Unless you want to also add it if `circular[seq_id]` exists but its value is empty.

[the remark this answers is not in the document]
Comment (MatBarba) [reply src/python/ensembl/brc4/runnable/integrity.py:545]: ok that double negative got me confused

[the remark this answers is not in the document]
Comment (vsitnik) [reply src/python/ensembl/brc4/runnable/integrity.py:545]: thanks. keeping as it is as this is a dictionary, not a set

File: src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm (hunk 1 above)
Comment (vsitnik) [reply src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm:423]: https://github.com/Ensembl/ensembl/blob/release/111/modules/Bio/EnsEMBL/Feature.pm#L1067 -- for me

Comment (vsitnik) [reply src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm:428]: added

File: src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm [hunk 3]
```diff
@@ -1136,15 +1260,60 @@ sub new_predicted_transcript {
   return $transcript;
 }
 
+
+sub circulise_coord {
```
[the remark this answers is not in the document]
Comment (vsitnik) [reply src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm]: done. thanks!

File: src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm [hunk 4]
```diff
@@ -684,6 +742,8 @@ sub infer_exons {
 }
 
 sub infer_translation {
+  # produce ($genomic_start, $genomic_end) to be used by translation_coordinates
+  # to infer offsets withing start / end exons
```
[the remark this answers is not in the document]
Comment (vsitnik) [reply src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm]: thanks!

File: src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm [hunk 5]
```diff
@@ -1136,15 +1260,60 @@ sub new_predicted_transcript {
...
+  my $exon_strand = $gff_exon->strand;
+
+  my $is_circular = $transcript->slice->is_circular(); 
+  my $slice_len = $transcript->slice->length(); 
+  my $slice_name= $transcript->slice->name(); 
+
+  my $msg = "exon $exon_id (start $exon_start, end $exon_end, strand $exon_strand) on slice $slice_name";
+  $self->log_throw("not a valid strand for $msg") if ($exon_strand != 1 && $exon_strand != -1);
+
+  # circulise only if seq_ergion start is not withing exon (exon[ | ]exon)
```
[the remark this answers is not in the document]
Comment (vsitnik) [reply src/perl/Bio/EnsEMBL/Pipeline/Runnable/EG/LoadGFF3/LoadGFF3.pm]: thanks!

**Pull request merged**